src/abacus-ts/websocket/lib/indexerWebsocket.ts

```diff
@@ -9,29 +10,89 @@ import { isTruthy } from '@/lib/isTruthy';
 import { ReconnectingWebSocket } from './reconnectingWebsocket';
 
 const NO_ID_SPECIAL_STRING_ID = '______EMPTY_ID______';
-const CHANNEL_ID_SAFE_DIVIDER = '//////////';
 const CHANNEL_RETRY_COOLDOWN_MS = timeUnits.minute;
 
+interface SubscriptionHandlerInput {
+  channel: string;
+  id: string | undefined;
+  batched: boolean;
+  handleBaseData: (data: any, fullMessage: any) => void;
+  handleUpdates: (updates: any[], fullMessage: any) => void;
+}
+
+type SubscriptionHandlerTrackingMetadata = {
+  receivedBaseData: boolean;
+  sentSubMessage: boolean;
+  // using subs for this data means we lose it when the user fully unsubscribes and thus might actually retry more often than expected
+  // but this is fine since every consumer has subs wrapped in resource managers that prevent lots of churn
+  lastRetryBecauseErrorMs: number | undefined;
+  lastRetryBecauseDuplicateMs: number | undefined;
+};
+type SubscriptionHandler = SubscriptionHandlerInput & SubscriptionHandlerTrackingMetadata;
+
+type SubscriptionMap = {
+  [channel: string]: {
+    [idOrSpecialString: string]: SubscriptionHandler;
+  };
+};
+
+class SubscriptionManager {
+  private subscriptions: SubscriptionMap = {};
+
+  hasSubscription(channel: string, id?: string): boolean {
+    const normalizedId = id ?? NO_ID_SPECIAL_STRING_ID;
+    return Boolean(this.subscriptions[channel]?.[normalizedId]);
+  }
+
+  getSubscription(channel: string, id?: string): SubscriptionHandler | undefined {
+    const normalizedId = id ?? NO_ID_SPECIAL_STRING_ID;
+    if (!this.hasSubscription(channel, normalizedId)) {
+      return undefined;
+    }
+    return this.subscriptions[channel]?.[normalizedId];
+  }
+
+  addSubscription(handler: SubscriptionHandler): boolean {
+    const normalizedId = handler.id ?? NO_ID_SPECIAL_STRING_ID;
+    const channel = handler.channel;
+
+    if (this.hasSubscription(channel, normalizedId)) {
+      return false;
+    }
+
+    this.subscriptions[channel] ??= {};
+    this.subscriptions[channel][normalizedId] = handler;
+    return true;
+  }
+
+  removeSubscription(channel: string, id: string | undefined): boolean {
+    const normalizedId = id ?? NO_ID_SPECIAL_STRING_ID;
+    if (!this.hasSubscription(channel, normalizedId)) {
+      return false;
+    }
+    delete this.subscriptions[channel]![normalizedId];
+    return true;
+  }
+
+  getAllSubscriptions(): SubscriptionHandler[] {
+    return Object.values(this.subscriptions)
+      .filter(isTruthy)
+      .flatMap((channelSubs) => Object.values(channelSubs))
+      .filter(isTruthy);
```

Comment:
is it possible to reduce to just the last filter? can just check for channelSub existance and return undefined prior